File: platform/polyglot/healthz_diagnostics/mjs/src/logger.mjs
/**
 * Logger module for healthz-diagnostics package.
 *
 * Provides defensive programming logging with consistent format:
 * [{LEVEL}] [{package}:{filename}] {message}
 *
 * Usage:
 *     import { create as createLogger } from './logger.mjs';
 *     const logger = createLogger("healthz-diagnostics", "config_store.mjs");
 *     logger.info("Operation completed");
 */

// Log levels
export const DEBUG = 10;
export const INFO = 20;
export const WARN = 30;
export const ERROR = 40;

// Default log level (can be overridden via environment variable)
const DEFAULT_LEVEL = INFO;
const LOG_LEVEL = parseInt(process.env.HEALTHZ_DIAGNOSTICS_LOG_LEVEL || String(DEFAULT_LEVEL), 10);

/**
 * Extract clean filename from path or import.meta.url
 * @param {string} filename - Raw filename or URL
 * @returns {string} Clean filename without extension
 */
function extractFilename(filename) {
    // Handle import.meta.url style (file:///path/to/file.mjs)
    if (filename.startsWith("file://")) {
        filename = filename.split("/").pop();
    }
    // Handle path style
    if (filename.includes("/")) {
        filename = filename.split("/").pop();
    }
    // Remove .mjs/.js extension
    if (filename.endsWith(".mjs")) {
        filename = filename.slice(0, -4);
    } else if (filename.endsWith(".js")) {
        filename = filename.slice(0, -3);
    }
    return filename;
}

/**
 * Logger class with package/filename context
 */
class Logger {
    #package;
    #filename;
    #level;
    #output;

    /**
     * @param {string} packageName - Package identifier
     * @param {string} filename - Source file identifier
     * @param {number} level - Log level threshold
     * @param {Function} output - Output function (default: console.log)
     */
    constructor(packageName, filename, level = LOG_LEVEL, output = console.log) {
        this.#package = packageName;
        this.#filename = extractFilename(filename);
        this.#level = level;
        this.#output = output;
    }

    /**
     * Internal log method with level check
     * @param {string} level - Level name
     * @param {number} levelValue - Numeric level value
     * @param {string} message - Log message
     */
    #log(level, levelValue, message) {
        if (levelValue >= this.#level) {
            const formatted = `[${level}] [${this.#package}:${this.#filename}] ${message}`;
            this.#output(formatted);
        }
    }

    /**
     * Log debug message (detailed internal state)
     * @param {string} message
     */
    debug(message) {
        this.#log("DEBUG", DEBUG, message);
    }

    /**
     * Log info message (normal operations)
     * @param {string} message
     */
    info(message) {
        this.#log("INFO", INFO, message);
    }

    /**
     * Log warning message (recoverable issues)
     * @param {string} message
     */
    warn(message) {
        this.#log("WARN", WARN, message);
    }

    /**
     * Log error message (failures requiring attention)
     * @param {string} message
     */
    error(message) {
        this.#log("ERROR", ERROR, message);
    }
}

/**
 * Create a logger instance for the given package and file.
 *
 * @param {string} packageName - Package identifier (e.g., "healthz-diagnostics")
 * @param {string} filename - Source file identifier (e.g., import.meta.url or "config_store.mjs")
 * @param {object} options - Optional configuration
 * @param {number} [options.level] - Log level override
 * @param {Function} [options.output] - Output function override
 * @returns {Logger} Logger instance
 *
 * @example
 * const logger = create("healthz-diagnostics", import.meta.url);
 * logger.info("Config loaded successfully");
 */
export function create(packageName, filename, options = {}) {
    const { level, output } = options;
    return new Logger(
        packageName,
        filename,
        level ?? LOG_LEVEL,
        output ?? console.log
    );
}

export { Logger };

File: platform/polyglot/healthz_diagnostics/py/healthz_diagnostics/logger.py
"""
Logger module for healthz_diagnostics package.

Provides defensive programming logging with consistent format:
[{LEVEL}] [{package}:{filename}] {message}

Usage:
    from .logger import create as create_logger
    logger = create_logger("healthz_diagnostics", __name__)
    logger.info("Operation completed")
"""

import os
from typing import Protocol, Optional, Callable

# Log levels
DEBUG = 10
INFO = 20
WARN = 30
ERROR = 40

# Default log level (can be overridden via environment variable)
_DEFAULT_LEVEL = INFO
_LOG_LEVEL = int(os.getenv("HEALTHZ_DIAGNOSTICS_LOG_LEVEL", _DEFAULT_LEVEL))


class LoggerProtocol(Protocol):
    """Protocol defining the logger interface."""
    def debug(self, message: str) -> None: ...
    def info(self, message: str) -> None: ...
    def warn(self, message: str) -> None: ...
    def error(self, message: str) -> None: ...


class Logger:
    """Logger instance with package/filename context."""

    def __init__(
        self,
        package_name: str,
        filename: str,
        level: int = _LOG_LEVEL,
        output: Optional[Callable[[str], None]] = None
    ):
        self._package = package_name
        self._filename = self._extract_filename(filename)
        self._level = level
        self._output = output or print

    def _extract_filename(self, filename: str) -> str:
        """Extract clean filename from module path."""
        # Handle __name__ style (e.g., "healthz_diagnostics.config_store")
        if "." in filename:
            filename = filename.split(".")[-1]
        # Remove .py extension if present
        if filename.endswith(".py"):
            filename = filename[:-3]
        return filename

    def _log(self, level: str, level_value: int, message: str) -> None:
        """Internal log method with level check."""
        if level_value >= self._level:
            formatted = f"[{level}] [{self._package}:{self._filename}] {message}"
            self._output(formatted)

    def debug(self, message: str) -> None:
        """Log debug message (detailed internal state)."""
        self._log("DEBUG", DEBUG, message)

    def info(self, message: str) -> None:
        """Log info message (normal operations)."""
        self._log("INFO", INFO, message)

    def warn(self, message: str) -> None:
        """Log warning message (recoverable issues)."""
        self._log("WARN", WARN, message)

    def error(self, message: str) -> None:
        """Log error message (failures requiring attention)."""
        self._log("ERROR", ERROR, message)


def create(
    package_name: str,
    filename: str,
    level: Optional[int] = None,
    output: Optional[Callable[[str], None]] = None
) -> Logger:
    """
    Create a logger instance for the given package and file.

    Args:
        package_name: Package identifier (e.g., "healthz_diagnostics")
        filename: Source file identifier (e.g., __name__)
        level: Optional log level override
        output: Optional output function override (default: print)

    Returns:
        Logger instance

    Example:
        logger = create("healthz_diagnostics", __name__)
        logger.info("Config loaded successfully")
    """
    return Logger(
        package_name=package_name,
        filename=filename,
        level=level if level is not None else _LOG_LEVEL,
        output=output
    )


# Expose log level constants for external configuration
__all__ = ["create", "Logger", "LoggerProtocol", "DEBUG", "INFO", "WARN", "ERROR"]
